File: research/hegelir/conformance/checker.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

// Conformance checker: read the golden vectors produced by the Rust
// reference and verify the Go port reproduces every output bit-for-bit.

func hx(s string) uint64 {
	v, err := strconv.ParseUint(s, 16, 64)
	if err != nil {
		panic(err)
	}
	return v
}

func main() {
	f, err := os.Open(os.Args[1])
	if err != nil {
		panic(err)
	}
	defer f.Close()

	var total, mismatch int
	byKind := map[string][2]int{} // kind -> {total, mismatch}
	var firstFails []string

	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 1<<20), 1<<20)
	for sc.Scan() {
		line := sc.Text()
		if line == "" {
			continue
		}
		p := strings.Split(line, "\t")
		kind := p[0]
		var got, want uint64
		var desc string
		switch kind {
		case "F2I":
			in, expect := hx(p[1]), hx(p[2])
			got = floatToIndex(math.Float64frombits(in))
			want = expect
			desc = fmt.Sprintf("F2I in=%016x", in)
		case "I2F":
			in, expect := hx(p[1]), hx(p[2])
			got = math.Float64bits(indexToFloat(in))
			want = expect
			desc = fmt.Sprintf("I2F in=%016x", in)
		case "SIR":
			lo, hi, expect := hx(p[1]), hx(p[2]), hx(p[3])
			got = math.Float64bits(simplestInRange(
				math.Float64frombits(lo), math.Float64frombits(hi)))
			want = expect
			desc = fmt.Sprintf("SIR lo=%016x hi=%016x", lo, hi)
		default:
			panic("unknown kind " + kind)
		}
		total++
		agg := byKind[kind]
		agg[0]++
		if got != want {
			mismatch++
			agg[1]++
			if len(firstFails) < 10 {
				firstFails = append(firstFails, fmt.Sprintf("  %s  got=%016x want=%016x", desc, got, want))
			}
		}
		byKind[kind] = agg
	}
	if err := sc.Err(); err != nil {
		panic(err)
	}

	fmt.Printf("checked %d vectors across %d kinds\n", total, len(byKind))
	for _, k := range []string{"F2I", "I2F", "SIR"} {
		if agg, ok := byKind[k]; ok {
			fmt.Printf("  %-4s %6d checked  %6d mismatched\n", k, agg[0], agg[1])
		}
	}
	if mismatch == 0 {
		fmt.Println("RESULT: PASS — Go port is bit-exact with the Rust reference")
	} else {
		fmt.Printf("RESULT: FAIL — %d/%d mismatched\n", mismatch, total)
		for _, s := range firstFails {
			fmt.Println(s)
		}
		os.Exit(1)
	}
}

File: research/hegelir/conformance/checker.rs
// Check the EMITTED Rust module against the Rust reference golden vectors.
// The generated functions are pulled in via include! so this compiles as one
// crate: rustc -O checker.rs (with float_index_gen.rs alongside).
#![allow(unused_parens, unused_variables, unused_mut, dead_code)]

use std::io::BufRead;

include!("float_index_gen.rs");

fn main() {
    let path = std::env::args().nth(1).expect("usage: check <vectors.tsv>");
    let file = std::fs::File::open(&path).unwrap();
    let reader = std::io::BufReader::new(file);

    let mut total: u64 = 0;
    let mut mism: u64 = 0;
    let mut counts: std::collections::BTreeMap<String, (u64, u64)> = std::collections::BTreeMap::new();
    let mut fails: Vec<String> = Vec::new();

    for line in reader.lines() {
        let line = line.unwrap();
        if line.is_empty() {
            continue;
        }
        let p: Vec<&str> = line.split('\t').collect();
        let kind = p[0];
        let (got, want, desc): (u64, u64, String) = match kind {
            "F2I" => {
                let inp = u64::from_str_radix(p[1], 16).unwrap();
                (
                    float_to_index(f64::from_bits(inp)),
                    u64::from_str_radix(p[2], 16).unwrap(),
                    format!("F2I in={}", p[1]),
                )
            }
            "I2F" => {
                let inp = u64::from_str_radix(p[1], 16).unwrap();
                (
                    index_to_float(inp).to_bits(),
                    u64::from_str_radix(p[2], 16).unwrap(),
                    format!("I2F in={}", p[1]),
                )
            }
            "SIR" => {
                let lo = u64::from_str_radix(p[1], 16).unwrap();
                let hi = u64::from_str_radix(p[2], 16).unwrap();
                (
                    simplest_in_range(f64::from_bits(lo), f64::from_bits(hi)).to_bits(),
                    u64::from_str_radix(p[3], 16).unwrap(),
                    format!("SIR lo={} hi={}", p[1], p[2]),
                )
            }
            _ => panic!("unknown kind {}", kind),
        };
        total += 1;
        let e = counts.entry(kind.to_string()).or_insert((0, 0));
        e.0 += 1;
        if got != want {
            mism += 1;
            e.1 += 1;
            if fails.len() < 10 {
                fails.push(format!("  {}  got={:016x} want={:016x}", desc, got, want));
            }
        }
    }

    println!("checked {} vectors across {} kinds", total, counts.len());
    for k in ["F2I", "I2F", "SIR"] {
        if let Some(v) = counts.get(k) {
            println!("  {:<4} {:6} checked  {:6} mismatched", k, v.0, v.1);
        }
    }
    if mism == 0 {
        println!("RESULT: PASS — emitted Rust is bit-exact with the Rust reference");
    } else {
        println!("RESULT: FAIL — {}/{} mismatched", mism, total);
        for s in &fails {
            println!("{}", s);
        }
        std::process::exit(1);
    }
}
